src/Compilers/Core/Portable/PEWriter/PeWriter.cs

```diff
@@ -88,10 +89,17 @@ internal sealed class PeWriter
             // If PDB writer is given, we have to have PDB path.
             Debug.Assert(nativePdbWriterOpt == null || pdbPathOpt != null);
 
-            var peWriter = new PeWriter(context.Module.Properties, context.Module.Win32Resources, context.Module.Win32ResourceSection, pdbPathOpt, deterministic);
-            var mdWriter = FullMetadataWriter.Create(context, messageProvider, allowMissingMethodBodies, deterministic, getPortablePdbStreamOpt != null, cancellationToken);
+            try
+            {
+                var peWriter = new PeWriter(context.Module.Properties, context.Module.Win32Resources, context.Module.Win32ResourceSection, pdbPathOpt, deterministic);
+                var mdWriter = FullMetadataWriter.Create(context, messageProvider, allowMissingMethodBodies, deterministic, getPortablePdbStreamOpt != null, cancellationToken);
 
-            return peWriter.WritePeToStream(mdWriter, getPeStream, getPortablePdbStreamOpt, nativePdbWriterOpt);
+                return peWriter.WritePeToStream(mdWriter, getPeStream, getPortablePdbStreamOpt, nativePdbWriterOpt);
+            }
+            catch (Exception ex) when (!(ex is PdbWritingException || ex is ResourceException || ex is PermissionSetFileReadException))
```

Comment:
> As I explained to @TyOverby I'd prefer the try catch to be in CommonCompiler around the call to Compilation.Emit

I disagree with this approach.  As previously stated all other IO exceptions are handled here and turned into diagnostics.  This propsoal would special case a very small subset of our IO.  This is not explainable to third parties.  The line would have to be:

> Calls to emit may throw IO exceptions or it may turn them into diagnostics.  Depends on when it happens.  

This change makes the change very simple to explain:

> Exceptions during IO will be represented as diagnostics.  

That is simple and explainable.  

> We should let exceptions thrown by user-code caused by incorrect implementation pass thru, 

That is not the case.  Everywhere else in the code base we blanket catch exceptions thrown by user code and turn it into diagonstics.  This is specifically to prevent bad user code from crashing the compiler.